src/util/permissions.tsx

```diff
@@ -0,0 +1,377 @@
+import { LxdIdentity, LxdPermission } from "types/permissions";
+import { OptionHTMLAttributes } from "react";
+import { LxdImage } from "types/image";
+import { useSupportedFeatures } from "context/useSupportedFeatures";
+import { useSettings } from "context/useSettings";
+
+export const defaultOption = {
+  disabled: true,
+  label: "Select an option",
+  value: "",
+};
+
+export const noneAvailableOption = {
+  disabled: true,
+  label: "None available",
+  value: "",
+};
+
+// the resource types comes from the openFGA authorisation model in lxd
+// ref: https://discourse.ubuntu.com/t/identity-and-access-management-for-lxd/41516
+export const resourceTypeOptions = [
+  {
+    ...defaultOption,
+  },
+  {
+    value: "server",
+    label: "Server",
+  },
+  {
+    value: "identity",
+    label: "Identity",
+  },
+  {
+    value: "group",
+    label: "Group",
+  },
+  {
+    value: "certificate",
+    label: "Certificate",
+  },
+  {
+    value: "project",
+    label: "Project",
+  },
+  {
+    value: "profile",
+    label: "Profile",
+  },
+  {
+    value: "instance",
+    label: "Instance",
+  },
+  {
+    value: "image",
+    label: "Image",
+  },
+  {
+    value: "image_alias",
+    label: "Image alias",
+  },
+  {
+    value: "storage_pool",
+    label: "Storage pool",
+  },
+  {
+    value: "storage_volume",
+    label: "Storage volume",
+  },
+  {
+    value: "storage_bucket",
+    label: "Storage bucket",
+  },
+  {
+    value: "network",
+    label: "Network",
+  },
+  {
+    value: "network_acl",
+    label: "Network ACL",
+  },
+  {
+    value: "network_zone",
+    label: "Network zone",
+  },
+];
+
+const sortOptions = (
+  a: OptionHTMLAttributes<HTMLOptionElement>,
+  b: OptionHTMLAttributes<HTMLOptionElement>,
+) => {
+  if (b.label === "Select an option") {
+    return 1;
+  }
+
+  return (a.label ?? "").localeCompare(b.label as string);
+};
+
+export const generateResourceOptions = (args: {
+  resourceType: string;
+  permissions?: LxdPermission[];
+  imageNamesLookup: Record<string, string>;
+  identityNamesLookup: Record<string, string>;
+}) => {
+  const { resourceType, permissions, identityNamesLookup, imageNamesLookup } =
+    args;
```

Comment:
Why not use plain arguments?

Reply:
If a function has over two parameters then I prefer using object inputs as positional arguments can get messy, especially when there is a mixture of optional and required arguments. wdyt?

Reply:
I am not a fan of the style as it mostly adds extra lines to unpack the arguments. A function should ideally not receive more than 5 parameters, and for that it seems more straight forward to directly pass the arguments.

Reply:
Yes good point, if a function starts to have messy interface it's probably time to refactor it anyway :+1: will update